File: aoj/DPL_1_A/a.cpp
#include <iostream>
#include <vector>

using namespace std;

int N, M;
vector<int> denom;
int dp[50001];
#define INF 99999999

int main()
{
    cin >> N >> M;
    fill( dp, dp+50001, INF );
    for( int i=0; i<M; i++ ){
        int c;
        cin >> c;
        denom.push_back(c);
    }

    dp[0] = 0;

    for( int i=1; i<=N; i++ ){
        for( int j=0; j<M; j++ ){
            if( denom[j] <= i ){
                dp[i] = min( dp[i-denom[j]]+1, dp[i] );
            }
        }
    }

    cout << dp[N] << endl;

    
}

File: aoj/DPL_1_A/a.rb
#!ruby

INF = 999999

coins = []

dp = Array.new( 100000, INF )

yen, denom = gets.split.map{|e| e.to_i}

denom.times do
  c = gets.to_i
  coins << c
end

dp[0] = 0

(0..yen).each do |i|
  coins.each do |c|
    if( i >= c )
      dp[i] = [ dp[i], dp[i-c]+1 ].min
    end
  end
end

puts dp[yen]

      
